starknet_py/net/schemas/rpc.py

```diff
@@ -510,3 +522,171 @@ class PendingTransactionsSchema(Schema):
+            return FunctionInvocation(**data["function_invocation"])
+        return {"revert_reason": data["revert_reason"]}
+
+
+class InvokeTransactionTraceSchema(Schema):
+    validate_invocation = fields.Nested(
+        FunctionInvocationSchema(), data_key="validate_invocation", load_default=None
+    )
+    execute_invocation = fields.Nested(
+        ExecuteInvocationSchema(), data_key="execute_invocation", required=True
+    )
+    fee_transfer_invocation = fields.Nested(
+        FunctionInvocationSchema(),
+        data_key="fee_transfer_invocation",
+        load_default=None,
+    )
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> InvokeTransactionTrace:
+        return InvokeTransactionTrace(**data)
+
+
+class DeclareTransactionTraceSchema(Schema):
+    validate_invocation = fields.Nested(
+        FunctionInvocationSchema(), data_key="validate_invocation", load_default=None
+    )
+    fee_transfer_invocation = fields.Nested(
+        FunctionInvocationSchema(),
+        data_key="fee_transfer_invocation",
+        load_default=None,
+    )
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> DeclareTransactionTrace:
+        return DeclareTransactionTrace(**data)
+
+
+class DeployAccountTransactionTraceSchema(Schema):
+    validate_invocation = fields.Nested(
+        FunctionInvocationSchema(), data_key="validate_invocation", load_default=None
+    )
+    constructor_invocation = fields.Nested(
+        FunctionInvocationSchema(), data_key="constructor_invocation", required=True
+    )
+    fee_transfer_invocation = fields.Nested(
+        FunctionInvocationSchema(),
+        data_key="fee_transfer_invocation",
+        load_default=None,
+    )
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> DeployAccountTransactionTrace:
+        return DeployAccountTransactionTrace(**data)
+
+
+class L1HandlerTransactionTraceSchema(Schema):
+    function_invocation = fields.Nested(
+        FunctionInvocationSchema(), data_key="function_invocation", required=True
+    )
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> L1HandlerTransactionTrace:
+        return L1HandlerTransactionTrace(**data)
+
+
+class TransactionTraceSchema(OneOfSchema):
+    type_schemas = {
+        "INVOKE": InvokeTransactionTraceSchema(),
+        "DECLARE": DeclareTransactionTraceSchema(),
+        "DEPLOY_ACCOUNT": DeployAccountTransactionTraceSchema(),
+        "L1_HANDLER": L1HandlerTransactionTraceSchema(),
+    }
+
+    def get_data_type(self, data):
+        # All possible types of transaction trace (loaded by sketchy logic),
+        # it's possible that more are added later in the future, and it needs to be reformatted
+        if "function_invocation" in data:
+            return "L1_HANDLER"
+        if "constructor_invocation" in data:
+            return "DEPLOY_ACCOUNT"
+        if "execute_invocation" in data:
+            return "INVOKE"
+        return "DECLARE"
```

Comment:
Discussed offline previously, but it should be mentioned that until [changes from this PR](https://github.com/starkware-libs/starknet-specs/pull/131) are added to nodes, this is unstable and might or might not fly depending on the node, since none of the properties are actually required in the spec. 

I'm not sure there's a good solution for this aside from waiting, unless someone has an idea.

Either way, we should definitely rework this once 0.5 drops, so maybe we should add an issue for that?

Reply:
Yup, it will need a refactor, but until actual 0.5.0 drop, it doesn't need to be touched. Issue [here](https://github.com/software-mansion/starknet.py/issues/1177)